errors: fix wrapped multierr behavior

ruseinov opened the pull request:
Fixes #713 #trivial

husio commented:
Does this really fix #713? This is a bug fix, but it does not touch any extension code.

ruseinov commented:
@husio The PR is a draft one.

ruseinov commented:
@husio @ethanfrey 
So I have introduced the changes to cash/coin, I have removed some extra wraps that might have affected how informative it all is, but otherwise it's even uglier.

I have found out what I have been suspecting in the first place - IMHO in case of coins it's actually nice to fail fast, because otherwise we have to iterate over and over for no good reason. 

Also, it looks like in some cases the if/else logic gets more complicated because we continue instead of returning. 

In terms of how we test things - it's all the same, nothing has really changed. So the only actual benefit of this approach is a nice error message that contains all the errors that occurred. But again, in case of cash or coin some of these messages are redundant.

ruseinov commented:
<!--
  1 Error: Please include a CHANGELOG ent...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ruseinov commented:
I would suggest adding an optional wrap argument to append to avoid ugly and verbose wraps.

husio commented on coin/coin.go:
```diff
@@ -260,22 +260,23 @@ func (c *Coin) Clone() *Coin {
+	var err error
 	if !IsCC(c.Ticker) {
-		return errors.Wrapf(errors.ErrCurrency, "invalid currency: %s", c.Ticker)
+		err = errors.Wrapf(errors.ErrCurrency, "invalid currency: %s", c.Ticker)
```
I think it's better to use `errors.Append` here as well for consistency. If someone will add an error before this check, it can cause a bug.

ruseinov replied:
yeah, I was thinking about this, but at the same time some tests would fail because of that too! 
I'm still waiting to see if this even makes sense to merge tbh, as I'm not happy with the fact that we can't tell which field is empty in a test when we receive `ErrIsEmpty`, for example.

ethanfrey replied:
> I'm still waiting to see if this even makes sense to merge tbh, as I'm not happy with the fact that we can't tell which field is empty in a test when we receive ErrIsEmpty, for example.

We couldn't see that before either.
Now, at least we see two fields are empty and a coin is mis-encoded.
And ideally if we read the logs, they tell us (a developer) which field is messed up. Sure, we cannot *programatically* determine fields, but since we usually Wrap with the field name in validation, this should be good enough for us to fix our code faster.

ruseinov commented:
Or even an extra api call that includes a wrap, something like Appendf.

husio commented:
> In terms of how we test things - it's all the same, nothing has really changed.

I totally agree with this. For me, better testing was the second reason for introducing groups of errors.

ruseinov commented:
> I totally agree with this. For me, better testing was the second reason for introducing groups of errors.

Yeah, not having fielded errors kind of makes the usefulness of all this very limited, but it'd be nice to see how this works for batch though. 

There is one more point to consider - and this is when we have several different errors with the same cause - we have absolutely no way to differentiate which of them failed. Unless we start checking error messages for substrings, which gets us back to where we started some time ago.

ruseinov commented:
@ethanfrey @husio @alpe 
I'd like to make several points to discuss before we merge this:
1. As you can see I have removed some wraps that might have contained useful info. I did this on purpose, cause it makes the code less verbose/ugly. I'd propose to add an `Appendf` method that would auto-wrap. Does that make sense?
2. Not being able to test errors by field makes it impossible to say which field returned an error, if we have several of them returning, say, `ErrEmpty`.

ethanfrey commented:
> Not being able to test errors by field makes it impossible to say which field returned an error, if we have several of them returning, say, ErrEmpty.

Repeating inline comment here as I think it is important:

We couldn't see that before either.
Now, at least we see two fields are empty and a coin is mis-encoded.
And ideally if we read the logs, they tell us (a developer) which field is messed up. Sure, we cannot programatically determine fields, but since we usually Wrap with the field name in validation, this should be good enough for us to fix our code faster.

I assume when we "%v" or "%+v" a multi-error we get a human-readable list of all the sub-errors, and each sub-error is wrapped with some good text. This is a great boon for human debugging and testing already. Having this work seemlessly with the UI is a few steps away and we really need to have @webmaster128 and @apanizo involved before making API designs that are intended for good frontend handling. For testing work, we have resolved the issue.

ethanfrey commented:
I like the idea of `Appendf`:

```go
func Appendf(existing error, toAdd error, msg string, args ...interface{}) error {
    return Append(existing, Wrapf(toAdd, msg, args...))
}
```

Correct? Would make usage nicer

ethanfrey commented on errors/multi_test.go:
```diff
@@ -57,6 +57,31 @@ func TestAddToMultiErr(t *testing.T) {
+		exp *Error
+	}{
+		"A wrapped multierr should pass an is check": {
+			err: Wrap(Append(Append(myErrNotFound, myErrState), myErrMsg), "wrapped"),
```
What about the case:
`Append(Wrap(Append(Wrap(`. This is two levels and the default behavior with proposed `Appendf`. Let's make sure it works well. Note the Wrap *between* Appends, which I would suspect *could* cause trouble unless properly coded. Let's make sure it works properly now and in the face of future changes

ruseinov replied:
Sure thing

ethanfrey commented on x/cash/msg.go:
```diff
@@ -30,26 +30,22 @@ func (SendMsg) Path() string {
+	} else {
+		err = errors.Append(err, s.Amount.Validate())
 	}
+	err = errors.Append(err, s.Src.Validate())
```
We lost useful info here "dst" vs "src".
Now we will get some invalid address message, but no longer even a human-readable message about which address is incorrect.
I think Appendf/AppendWrap would encourage including this info

ethanfrey commented on x/cash/msg.go:
```diff
@@ -30,26 +30,22 @@ func (SendMsg) Path() string {
+		err = errors.Append(err, s.Amount.Validate())
 	}
+	err = errors.Append(err, s.Src.Validate())
+	err = errors.Append(err, s.Dest.Validate())
```
In other news, I think the use of Append here is actually cleaner than the if/return logic.
Just make sure Append and Wrap pass `nil` input errors through without changes, and it just flows

ethanfrey commented on x/cash/msg.go:
```diff
@@ -91,47 +87,45 @@ func (f *FeeInfo) DefaultPayer(addr []byte) *FeeInfo {
+	} else {
+		err = errors.Append(err, fee.Validate())
+
+		if !fee.IsNonNegative() {
```
Interesting, we need to update checks, as we relied on short-circuit in original validation.

ethanfrey commented on x/cash/msg.go:
```diff
@@ -91,47 +87,45 @@ func (f *FeeInfo) DefaultPayer(addr []byte) *FeeInfo {
-		if err := c.Owner.Validate(); err != nil {
-			return errors.Wrap(err, "owner address")
-		}
+		err = c.Owner.Validate()
```
Please don't remove Wraps. They were there for a good purpose.

We had tagged almost every validation message with a human-readable text to identify the field with errors to help debugging tests at least. Removing them now leads to the exact problem you were complaining about, having no field mappings.

Please leave them in now. We could replace it with `errors.Tag` instead of `errors.Wrap` in some future PR design possibly, but don't remove the info without an alternative.

Thinking about this, I think Appendf is a bad name, as we might want for Wrap or Tag...

AppendWrap, AppendWrapf, AppendTag as shortcuts???
Someone who likes simple APIs is gonna hate me for that proposal...

husio commented:
Keep in mind that stack trace produced by [proposed `Appendf`](https://github.com/iov-one/weave/pull/727#issuecomment-499070529) will capture the wrong frame as the `errors.Wrap` doesn't know it's being wrapped.

ethanfrey commented:
> Keep in mind that stack trace produced by proposed Appendf will capture the wrong frame as the errors.Wrap doesn't know it's being wrapped.

Good point.
Let's just do the verbose (but correct) approach everywhere in this PR.
Take the time to clean it up later.

At least having one example exposed on send tx, we can start getting feedback from consumers on how they parse this json, if it is useful info, etc.
